MOBILE-2277: Implement serializable w_module

brianblanchard-wf opened the pull request:
## Description
- Added a way to serialize w_modules to allow for communication with other systems (such as a mobile app)

## What was changed
- A separate entry point for a serializable module so all the existing w_module logic does not have any browser specific dependencies.
- Added a `SerializableBus` singleton that registers for events and listens for api calls to modules
- Added a `SerializableModule` class that extends the existing `Module` class that adds needed information to serialize a module
- Added a `Reflectable` class that allows classes and methods to be reflected upon using mirrors.
- Added a `SerializableEvent` and `SerializableEvents` class that allows events to be serializable.
- Unit tests around additions and changes.

Please review: @trentgrover-wf @evanweible-wf @maxwellpeterson-wf 

An actual implementation of these changes can be seen in this [diff](https://github.com/Workiva/w_viewer/compare/master...brianblanchard-wf:MOBILE-2281)

aviary2-wf commented:
## Raven
Number of Findings: 0

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,237 @@
+      }
+    } else {
+      _logger
+          .warning('Events not defined for ${module.serializableKey} module');
```
I think `info` would be more appropriate here - modules aren't required to define any events.

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,237 @@
+    dataToSend['event'] = eventKey;
+
+    if (data is JsonSerializable) {
+      data = (data as JsonSerializable).toJson();
```
You should be able to avoid the explicit `as` type cast here since you know it is of type `JsonSerializable` via the `if` check.

brianblanchard-wf replied:
Without the explicit cast, there is an analyzer error.

evanweible-wf replied:
Okay, it's fine as is then. Out of curiosity - what version of the Dart SDK are you on?

brianblanchard-wf replied:
1.19.1

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,237 @@
+    String method = apiCall['method'];
+    Object data = apiCall["data"];
+
+    SerializableModule targetModule = _registeredModules[module];
```
null check this

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,237 @@
+  void _deserializeAndCall(
+      SerializableModule module, String method, List data) {
+    InstanceMirror apiMirror = reflect(module.api);
+    ClassMirror classMirror = apiMirror.type;
```
null check on `apiMirror`

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,237 @@
+      SerializableModule module, String method, List data) {
+    InstanceMirror apiMirror = reflect(module.api);
+    ClassMirror classMirror = apiMirror.type;
+    MethodMirror apiMethodMirror = classMirror.declarations[new Symbol(method)];
```
null check on `classMirror`

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    // Check here that the position args in data match the expected params of the method being called
+    if (data is List && apiMethodMirror.parameters.length == data.length) {
```
The method signature already requires that `data` is a `List`, so this first part should never be false

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,237 @@
+
+        // If the type data being passed to this param is not equal to the expected type
+        // try to serialize it into an Dart class
+        if (param.type.reflectedType != data[i].runtimeType) {
```
I'm not very familiar with mirrors at all - how does this handle generics? For example, if the reflected param type is `List<int>` but the data runtime type could only be inferred to be `List` but in practice only contained `int`s, would that be an issue?

brianblanchard-wf replied:
It will work. After some testing I had to make some changes to how this works. That `if (param.type.reflectedType != data[i].runtimeType` check does not alway work as expected. For example, when passing an array from JS, the type is `JsArray` but if you do a type check on like `jsArray is List` that returns true. So now it checks to see if data being passed is a `Map` and then will try to serialize it into a the expected object of the api method.

I know this is not idea, but before this is used at any type of scale, all this serialization stuff will be replaced by frugal so I think it is ok for now.

evanweible-wf replied:
Yup, no problem, was mainly just curious

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,237 @@
+              return;
+            }
+          } else {
+            _logger.warning('Incompatiable type for deserialization');
```
Maybe add some more context here (module key, event key, expected type vs. actual incompatible type)

brianblanchard-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,237 @@
+}
+
+// This class is used to communicate across a WKWebView in an iOS app
+class WKWebViewBridge extends Bridge<CustomEvent> {
```
Where do you think this should live? I was told to keep it in here for now.

evanweible-wf replied:
I think it can live with the application you're working on until it has been vetted and at that point it could probably live in wdesk_sdk or something similar

brianblanchard-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,237 @@
+
+// Any classes / methods that are going to be reflected must annotate with this
+class Reflectable {
+  const Reflectable();
```
No you have to define your own class you annotate with. https://api.dartlang.org/dev/1.20.0-dev.9.0/dart-mirrors/MirrorsUsed/metaTargets.html

brianblanchard-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,237 @@
+
+  JsObject get wkBridge {
+    if (context['webkit'] != null) {
+      return context['webkit']['messageHandlers']['bridge'];
```
If the `webkit` is there then the `messageHandlers` will always be there. Also, when the `wkBridge` is invoked using an optional `wkBridge?.callMethod('postMessage', [new JsObject.jsify(dataToSend)]); `

brianblanchard-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,237 @@
+      : super(dispatchKey);
+
+  String _eventKey;
+  String get eventKey => _eventKey;
```
That is a problem we will have to tackle in the future, being able to register a unique instance with bridge. When I talked with Trent initially we decided for now it is ok to just be able to handle one instance of a given module at a time.

brianblanchard-wf commented:
@evanweible-wf thanks for all the great feedback. I believe I addressed all your concerns. Please take another look when you have time.

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,224 @@
+
+  Bridge get bridge => _bridge;
+  Map<String, _ModuleRegistration> get moduleRegistrations =>
+      _moduleRegistrations;
```
I'm not sure it makes sense to expose the `_ModuleRegistration` instances, I see them as being implementation details that make disposal easier. It think exposing the modules like you had before would be fine:

```dart
Map<String, SerializableModule> get registeredModules
    => new Map.fromIterable(_moduleRegistrations.keys, value: (key) => _moduleRegistrations[key].module);
```

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,224 @@
+  Map<String, _ModuleRegistration> get moduleRegistrations =>
+      _moduleRegistrations;
+
+  set bridge(Bridge bridge) {
```
If a field has a getter and a setter, they should be grouped together

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,224 @@
+        if (apiMethodMirror == null) {
+          _logger.warning(
+              'Method $method does not exist on ${module.serializableKey}\' module\'s API');
+          return;
```
I like this approach of logging/returning early in the failure case instead of nested ifs, could you do this for the `apiMirror` and `classMirror` null checks above, too?

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,224 @@
+    if (apiMirror != null) {
+      ClassMirror classMirror = apiMirror.type;
+
+      if (ClassMirror != null) {
```
s/`ClassMirror`/`classMirror`

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,224 @@
+    }
+  }
+
+  void _registerForAllEvents(SerializableModule module) {
```
you'll want to pass in the `_ModuleRegistration` instance here, too, so that you can have it manage the event subscriptions below

evanweible-wf commented on test/serializable_test.dart:
```diff
@@ -0,0 +1,333 @@
+
+    test('should provide a singleton', () async {
+      expect(SerializableBus.sharedBus, isNotNull);
+      expect(SerializableBus.sharedBus is SerializableBus, isTrue);
```
```dart
expect(SerializableBus.sharedBus, new isInstanceOf<SerializableBus>());
```

evanweible-wf commented on tool/dev.dart:
```diff
@@ -31,6 +31,7 @@ Future<Null> main(List<String> args) async {
   ];
   config.copyLicense.directories = directories;
   config.format.directories = directories;
+  config.test.platforms = ['content-shell', 'vm'];
```
If you switch to the individual w_common entry points above, I believe you'll be able to run these tests on the VM instead of in the browser

brianblanchard-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,224 @@
+      return;
+    }
+
+    ClassMirror classMirror = apiMirror.type;
```
A null check not needed here. If the `apiMirror` exists then a class mirror must also exist.

brianblanchard-wf commented:
@evanweible-wf I have addressed your last round of feedback. Please review.

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,224 @@
+
+@MirrorsUsed(metaTargets: 'serializable_module.src.serializable.Reflectable')
+import 'dart:mirrors';
+import 'dart:async';
```
#nit alphabetize imports

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,224 @@
+  SerializableEvent(this._eventKey, DispatchKey dispatchKey)
+      : super(dispatchKey);
+
+  String get eventKey => _eventKey;
```
could simplify this to a final variable:

```dart
final String eventKey;
SerializableEvent(this.eventKey, DispatchKey dispatchKey) : super(dispatchKey);
```

evanweible-wf replied:
could simplify this to a final variable:

```dart
final String eventKey;
SerializableEvent(this.eventKey, DispatchKey dispatchKey) : super(dispatchKey);
```

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    // Check here that the position args in data match the expected params of the method being called
+    if (apiMethodMirror.parameters.length == data.length) {
```
could also invert this if check here and return early if `apiMethodMirror.parameters.length != data.length`

evanweible-wf commented on test/serializable_test.dart:
```diff
@@ -0,0 +1,317 @@
+import 'package:w_common/json_serializable.dart' show JsonSerializable;
+import 'package:w_module/w_module.dart';
+import 'package:test/test.dart';
+import 'package:mockito/mockito.dart';
```
sort imports

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,224 @@
+  Bridge get bridge => _bridge;
+  set bridge(Bridge bridge) {
+    _bridge = bridge;
+    _bridge.apiCallReceived.listen(_handleApiCall);
```
Should store this `StreamSubscription` and then cancel it when a new bridge is set

evanweible-wf commented:
@Workiva/client-platform-pp @Workiva/ref-pp anyone else care to take a look?

dustinlessard-wf commented:
Looking - will post when done.

dustinlessard-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,224 @@
+
+  void reset() {
+    _moduleRegistrations.clear();
+    _bridge = null;
```
Looking at the [Bridge abstract class](https://github.com/Workiva/w_module/pull/59/files#diff-949ea1a52f226534b3058b4ce26a64edR33), and the [proposed implementation](https://github.com/Workiva/w_viewer_mobile/blob/MOBILE-2307/lib/src/wk_web_view_bridge.dart#L17) : 
We see that there is a `StreamSubscription` created yet never cancelled.

Looking at the reset method, we see that we simply null out the reference to _bridge.  

Two questions: 
1. Should `WKWebViewBridge` support a function to call when we want to clean up resources (aka dispose)?
2. Should `SerializableBus` call that cleanup function prior to nulling out the `_bridge` reference?

One option is to consider using w_common : https://github.com/Workiva/w_common/blob/master/lib/disposable.dart

dustinlessard-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,224 @@
+
+    event['data'] = data;
+
+    _bridge?.broadcastSerializedEvent(event);
```
Do we want to know if the `_bridge` is null at the time that we attempt to send an event?  If so, we may consider adding some logging around that.

brianblanchard-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,224 @@
+        }
+      }
+    } else {
+      _logger.info(
```
In an earlier comment @evanweible-wf suggested that `info` were to be used because what it is logging is not a required for a module to implement.

dustinlessard-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,224 @@
+          } on NoSuchMethodError {
+            _logger.warning(
+                '${paramClassMirror.simpleName.toString()} does not implement fromJson named constructor');
+            return;
```
nit: Is this needed?

brianblanchard-wf replied:
I think it is useful because we get a meaningful error message instead of the error being thrown in the console.

brianblanchard-wf commented:
@dustinlessard-wf @evanweible-wf I have addressed for latest feedback. I added the `Disposable` mixin to the `Bridge` abstract class that is now used to manage the stream subscriptions to the bridge.

dustinlessard-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,235 @@
+
+  Bridge get bridge => _bridge;
+  set bridge(Bridge bridge) {
+    _bridge = bridge;
```
Caveat: I'm uncertain of the answer to this question:
If `_bridge` is not null at the time we attempt to set it, should we `dispose` the previous one?

brianblanchard-wf replied:
I don't think it would hurt to do that.

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -30,20 +30,23 @@ class Reflectable {
   void handleSerializedApiCall(T apiCall);
+
+  @override
+  void manageStreamSubscription(StreamSubscription subscription) {
```
I think this should be private - by overriding it and omitting the `@protected` label, you're effectively allowing anyone who has access to the bridge to add subscriptions that should be managed by that bridge, when in reality that should only be a handled by the bridge itself. If you make this private, then you can still call it below without exposing that ability to external consumers.

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -82,8 +85,13 @@ class SerializableBus {
   set bridge(Bridge bridge) {
+    if (_bridge != null) {
+      _bridge.dispose();
+    }
```
`_bridge?.dispose()`

evanweible-wf commented on lib/src/serializable.dart:
```diff
@@ -82,8 +85,13 @@ class SerializableBus {
+
     _bridge = bridge;
-    _bridge.apiCallReceived.listen(_handleApiCall);
+    _bridge.manageStreamSubscription(
```
Probably need to ensure that `bridge` isn't being set to null here, otherwise this would throw (would also prevent a failure below when reset() is called)

maxwellpeterson-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,239 @@
+
+  @override
+  void manageStreamSubscription(StreamSubscription subscription) {
+    super.manageStreamSubscription(subscription);
```
Does this method need to be overridden if it is just calling super?

evanweible-wf replied:
This should probably be private and only used in this file, otherwise external consumers could tell the bridge to manage stream subscriptions that it shouldn't

maxwellpeterson-wf commented on lib/src/serializable.dart:
```diff
@@ -0,0 +1,239 @@
+
+  @override
+  void manageStreamSubscription(StreamSubscription subscription) {
+    super.manageStreamSubscription(subscription);
```
Same question about overriding here.

evanweible-wf commented:
+1

dustinlessard-wf commented:
Are we ok with decreasing coverage by nearly 7%?  Is that accurate?

brianblanchard-wf commented:
@dustinlessard-wf I have added some more unit tests. Is the level of coverage now acceptable?

dustinlessard-wf commented:
+1 - I think it was probably acceptable before, but I wasn't certain about how accurate CodeCov was being.  (I still have my doubts but this is totally great!)

evanweible-wf commented:
+1

sebastianmalysa-wf commented:
+1

maxwellpeterson-wf commented:
+1

jordanross-wf commented:
+10 from a functionality standpoint

maxwellpeterson-wf commented:
@jayudey-wf ready for you

jayudey-wf commented:
QA Resource Approval: +10

- [ ] Testing instruction (from offline discussion with brian)
- [ ] Dev +1's
- [ ] Dev/QA +10 with detail of what was tested
 - performed and documented by Jordan
- [ ] Unit test created/updated
- [ ] All unit tests pass

Merging into master.
